File: decoder.go
package codexcw

import (
	"encoding/json"
	"fmt"
	"time"
)

type wireEvent struct {
	Type     EventType       `json:"type"`
	ThreadID string          `json:"thread_id"`
	Usage    Usage           `json:"usage"`
	Item     json.RawMessage `json:"item"`
	Error    json.RawMessage `json:"error"`
	Message  string          `json:"message"`
}

type wireItem struct {
	ID                string       `json:"id"`
	Type              ItemType     `json:"type"`
	Status            string       `json:"status"`
	Text              string       `json:"text"`
	Message           string       `json:"message"`
	Command           string       `json:"command"`
	AggregatedOutput  string       `json:"aggregated_output"`
	ExitCode          *int         `json:"exit_code"`
	Changes           []FileChange `json:"changes"`
	Tool              string       `json:"tool"`
	SenderThreadID    string       `json:"sender_thread_id"`
	ReceiverThreadIDs []string     `json:"receiver_thread_ids"`
}

func decodeEvent(line []byte, runID string, threadID string, now time.Time) (Event, error) {
	raw := append(json.RawMessage(nil), line...)

	var wire wireEvent
	if err := json.Unmarshal(line, &wire); err != nil {
		return Event{}, err
	}
	if wire.Type == "" {
		return Event{}, fmt.Errorf("missing event type")
	}

	event := Event{
		Type:       wire.Type,
		RunID:      runID,
		ThreadID:   threadID,
		ReceivedAt: now,
		Raw:        raw,
	}

	switch wire.Type {
	case EventThreadStarted:
		event.ThreadID = wire.ThreadID
		event.ThreadStarted = &ThreadStartedEvent{ThreadID: wire.ThreadID}
	case EventTurnStarted:
		event.TurnStarted = &TurnStartedEvent{}
	case EventTurnCompleted:
		event.TurnCompleted = &TurnCompletedEvent{Usage: normalizeCodexUsage(wire.Usage)}
	case EventTurnFailed:
		event.TurnFailed = &TurnFailedEvent{
			Error: decodeEventError(wire.Error),
			Usage: normalizeCodexUsage(wire.Usage),
		}
	case EventItemStarted:
		item, err := decodeItem(wire.Item)
		if err != nil {
			return Event{}, err
		}
		event.ItemStarted = &ItemEvent{Item: item}
	case EventItemCompleted:
		item, err := decodeItem(wire.Item)
		if err != nil {
			return Event{}, err
		}
		event.ItemCompleted = &ItemEvent{Item: item}
	case EventError:
		event.Error = &CodexErrorEvent{Message: wire.Message, Raw: append(json.RawMessage(nil), wire.Error...)}
		if event.Error.Message == "" && len(wire.Error) > 0 {
			event.Error.Message = string(wire.Error)
		}
	}

	return event, nil
}

func decodeItem(raw json.RawMessage) (Item, error) {
	if len(raw) == 0 {
		return Item{}, fmt.Errorf("missing item payload")
	}
	var wire wireItem
	if err := json.Unmarshal(raw, &wire); err != nil {
		return Item{}, err
	}
	return Item{
		ID:                wire.ID,
		Type:              wire.Type,
		Status:            wire.Status,
		Raw:               append(json.RawMessage(nil), raw...),
		Text:              wire.Text,
		Message:           wire.Message,
		Command:           wire.Command,
		AggregatedOutput:  wire.AggregatedOutput,
		ExitCode:          wire.ExitCode,
		Changes:           wire.Changes,
		Tool:              wire.Tool,
		SenderThreadID:    wire.SenderThreadID,
		ReceiverThreadIDs: wire.ReceiverThreadIDs,
	}, nil
}

// normalizeCodexUsage derives the total when Codex omits total_tokens.
// cached_input_tokens is a subset of input_tokens on the codex wire, so the
// derived total is input plus output; an explicit total is preserved.
func normalizeCodexUsage(usage Usage) Usage {
	if usage.TotalTokens == 0 {
		usage.TotalTokens = usage.InputTokens + usage.OutputTokens
	}
	return usage
}

func decodeEventError(raw json.RawMessage) ErrorPayload {
	err := ErrorPayload{Raw: append(json.RawMessage(nil), raw...)}
	if len(raw) == 0 {
		return err
	}
	var wire struct {
		Message string `json:"message"`
	}
	if json.Unmarshal(raw, &wire) == nil {
		err.Message = wire.Message
	}
	if err.Message == "" {
		err.Message = string(raw)
	}
	return err
}

File: crates/codexcw/src/decoder.rs
//! JSONL line decoding into [`Event`] values.

use std::time::SystemTime;

use serde::Deserialize;
use serde_json::value::RawValue;

use crate::event::{
    CodexErrorEvent, ErrorPayload, Event, EventKind, EventPayload, FileChange, Item, ItemKind,
    Usage,
};

#[derive(Deserialize)]
struct WireEvent {
    #[serde(rename = "type")]
    kind: Option<String>,
    #[serde(default)]
    thread_id: String,
    #[serde(default)]
    usage: Usage,
    #[serde(default)]
    item: Option<Box<RawValue>>,
    #[serde(default)]
    error: Option<Box<RawValue>>,
    #[serde(default)]
    message: String,
}

#[derive(Deserialize)]
struct WireItem {
    #[serde(default)]
    id: String,
    #[serde(default, rename = "type")]
    kind: String,
    #[serde(default)]
    status: String,
    #[serde(default)]
    text: String,
    #[serde(default)]
    message: String,
    #[serde(default)]
    command: String,
    #[serde(default)]
    aggregated_output: String,
    #[serde(default)]
    exit_code: Option<i32>,
    #[serde(default)]
    changes: Vec<FileChange>,
}

/// Decodes one JSONL line into an [`Event`].
///
/// Returns a human-readable message on malformed input; the caller wraps it
/// into a line-numbered decode error.
pub(crate) fn decode_event(
    line: &[u8],
    run_id: &str,
    thread_id: &str,
    now: SystemTime,
) -> Result<Event, String> {
    let raw = String::from_utf8_lossy(line).into_owned();

    let wire: WireEvent = serde_json::from_slice(line).map_err(|err| err.to_string())?;
    let kind = wire.kind.unwrap_or_default();
    if kind.is_empty() {
        return Err("missing event type".to_string());
    }

    let mut event = Event {
        kind: EventKind::from_wire(&kind),
        run_id: run_id.to_string(),
        thread_id: thread_id.to_string(),
        received_at: now,
        raw,
        payload: EventPayload::Other,
    };

    event.payload = match event.kind {
        EventKind::ThreadStarted => {
            event.thread_id = wire.thread_id.clone();
            EventPayload::ThreadStarted {
                thread_id: wire.thread_id,
            }
        }
        EventKind::TurnStarted => EventPayload::TurnStarted,
        EventKind::TurnCompleted => EventPayload::TurnCompleted { usage: wire.usage },
        EventKind::TurnFailed => EventPayload::TurnFailed {
            error: decode_event_error(wire.error.as_deref()),
            usage: wire.usage,
        },
        EventKind::ItemStarted => EventPayload::ItemStarted(decode_item(wire.item.as_deref())?),
        EventKind::ItemCompleted => EventPayload::ItemCompleted(decode_item(wire.item.as_deref())?),
        EventKind::Error => {
            let raw_error = wire
                .error
                .as_deref()
                .map(|value| value.get().to_string())
                .unwrap_or_default();
            let mut message = wire.message;
            if message.is_empty() && !raw_error.is_empty() {
                message = raw_error.clone();
            }
            EventPayload::Error(CodexErrorEvent {
                message,
                raw: raw_error,
            })
        }
        EventKind::Other(_) => EventPayload::Other,
    };

    Ok(event)
}

fn decode_item(raw: Option<&RawValue>) -> Result<Item, String> {
    let raw = raw.ok_or_else(|| "missing item payload".to_string())?;
    let text = raw.get();
    if text.is_empty() {
        return Err("missing item payload".to_string());
    }
    let wire: WireItem = serde_json::from_str(text).map_err(|err| err.to_string())?;
    Ok(Item {
        id: wire.id,
        kind: ItemKind::from_wire(&wire.kind),
        status: wire.status,
        raw: text.to_string(),
        text: wire.text,
        message: wire.message,
        command: wire.command,
        aggregated_output: wire.aggregated_output,
        exit_code: wire.exit_code,
        changes: wire.changes,
    })
}

fn decode_event_error(raw: Option<&RawValue>) -> ErrorPayload {
    let Some(raw) = raw else {
        return ErrorPayload::default();
    };
    let text = raw.get().to_string();
    if text.is_empty() {
        return ErrorPayload::default();
    }

    #[derive(Deserialize)]
    struct Inner {
        #[serde(default)]
        message: String,
    }

    let mut message = serde_json::from_str::<Inner>(&text)
        .map(|inner| inner.message)
        .unwrap_or_default();
    if message.is_empty() {
        message = text.clone();
    }
    ErrorPayload { message, raw: text }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn decodes_thread_started_and_backfills_thread_id() {
        let event = decode_event(
            br#"{"type":"thread.started","thread_id":"t-1"}"#,
            "run-x",
            "",
            SystemTime::UNIX_EPOCH,
        )
        .unwrap();
        assert_eq!(event.kind, EventKind::ThreadStarted);
        assert_eq!(event.thread_id, "t-1");
        assert_eq!(event.run_id, "run-x");
    }

    #[test]
    fn decodes_agent_message_item() {
        let event = decode_event(
            br#"{"type":"item.completed","item":{"id":"i0","type":"agent_message","text":"Oi."}}"#,
            "run-x",
            "t-1",
            SystemTime::UNIX_EPOCH,
        )
        .unwrap();
        let item = event.item_completed().unwrap();
        assert_eq!(item.kind, ItemKind::AgentMessage);
        assert_eq!(item.text, "Oi.");
        assert_eq!(event.thread_id, "t-1");
    }

    #[test]
    fn decodes_command_execution_exit_code() {
        let event = decode_event(
            br#"{"type":"item.completed","item":{"id":"i0","type":"command_execution","command":"false","exit_code":7,"status":"failed"}}"#,
            "run-x",
            "",
            SystemTime::UNIX_EPOCH,
        )
        .unwrap();
        let item = event.item_completed().unwrap();
        assert_eq!(item.exit_code, Some(7));
        assert_eq!(item.status, "failed");
    }

    #[test]
    fn decodes_collab_tool_call_kind() {
        let event = decode_event(
            br#"{"type":"item.started","item":{"id":"i0","type":"collab_tool_call","tool":"wait","sender_thread_id":"t-parent","receiver_thread_ids":[],"agents_states":{},"status":"in_progress"}}"#,
            "run-x",
            "t-parent",
            SystemTime::UNIX_EPOCH,
        )
        .unwrap();
        let item = event.item_started().unwrap();
        assert_eq!(item.kind, ItemKind::CollabToolCall);
        assert_eq!(item.status, "in_progress");
        assert!(item.raw.contains("\"tool\":\"wait\""));
    }

    #[test]
    fn missing_type_is_an_error() {
        let err = decode_event(b"{}", "run", "", SystemTime::UNIX_EPOCH).unwrap_err();
        assert_eq!(err, "missing event type");
    }

    #[test]
    fn invalid_json_is_an_error() {
        let err = decode_event(b"not-json", "run", "", SystemTime::UNIX_EPOCH).unwrap_err();
        assert!(!err.is_empty());
    }

    #[test]
    fn turn_failed_error_message() {
        let event = decode_event(
            br#"{"type":"turn.failed","error":{"message":"turn broke"}}"#,
            "run",
            "",
            SystemTime::UNIX_EPOCH,
        )
        .unwrap();
        match event.payload {
            EventPayload::TurnFailed { error, .. } => assert_eq!(error.message, "turn broke"),
            other => panic!("unexpected payload: {other:?}"),
        }
    }
}
